File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> map = new HashMap<>();
        int res = 0;
        int n = s.length();
        
        for(int i = 0, l = -1; i < n; i++) {
            char c = s.charAt(i);
            int last = map.getOrDefault(c, -1);
            l = Math.max(l, last);
            res = Math.max(res, i - l);
            map.put(c, i);
        }
        
        return res;
    }
}

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.scala
import scala.collection.mutable._

object Solution {
    def lengthOfLongestSubstring(s: String): Int = {
        var res = 0
        var p = -1
        val map: Map[Char, Int] = Map()
        
        for(i <- 0 to s.length - 1) {
            val c = s.charAt(i)
            p = Math.max(p, map.get(c).getOrElse(-1))
            res = Math.max(res, i - p)
            map += (c -> i)
        }
        
        res
    }
}